runtime/ints.c

```diff
@@ -280,6 +280,157 @@ CAMLprim value caml_int32_shift_right(value v1, value v2)
 CAMLprim value caml_int32_shift_right_unsigned(value v1, value v2)
 { return caml_copy_int32((uint32_t)Int32_val(v1) >> Int_val(v2)); }
 
+#if defined(__GNUC__)
+#if ARCH_INT32_TYPE == long
+#define int32_clz __builtin_clzl
+#define int32_popcnt __builtin_popcountl
+#else /* ARCH_INT32_TYPE == long */
+#define int32_clz __builtin_clz
+#define int32_popcnt __builtin_popcount
+#endif /* ARCH_INT32_TYPE == long */
+
+#define int64_clz __builtin_clzll
+#define int64_popcnt __builtin_popcountll
+
+#else /* defined(__GNUC__) */
+ /* _MSC_ does not have builtin for popcnt, and
+  *  has undefined behavior for clz builtin when
+  *  lzcnt instruction is not supported.
+  *  Use naive version of clz and popcnt from Hacker's Delight. */
+
+int naive_int64_clz(uint64_t v)
+{
+  int n = 0;
+  int64_t x = (int64_t) v;
+  int64_t y = x;
+
+L: if (x < 0) return n;
+   if (y == 0) return 64 - n;
+   n = n + 1;
+   x = x << 1;
+   y = y >> 1;
+   goto L;
```

Comment:
Why use `goto` here rather than structured control flow constructs?

Reply:
It's from Hacker's Delight. I can rewrite it as a loop, or perhaps it would be better to replace with MSVC intrinsic call as suggested above.